crates/aptos/src/op/key.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    common::{
+        types::{CliCommand, EncodingType, Error, KeyType},
+        utils::{prompt_yes, to_common_success_result},
+    },
+    CliResult,
+};
+use aptos_crypto::{ed25519::Ed25519PrivateKey, x25519, Uniform, ValidCryptoMaterial};
+use rand::SeedableRng;
+use std::{fs::File, io::Write, path::PathBuf, process::exit};
+use structopt::StructOpt;
+
+#[derive(Debug, StructOpt)]
+pub enum KeyTool {
+    Generate(GenerateKey),
+}
+
+impl KeyTool {
+    pub async fn execute(self) -> CliResult {
+        match self {
+            KeyTool::Generate(generate) => generate.execute(),
+        }
+    }
+}
+
+#[derive(Debug, StructOpt)]
+pub struct GenerateKey {
+    /// Key type `x25519` or `ed25519`
+    #[structopt(short = "t", long, default_value = "ed25519")]
+    key_type: KeyType,
+    #[structopt(flatten)]
+    save_params: SaveKey,
+}
+
+impl CliCommand for GenerateKey {
+    fn execute(self) -> CliResult {
+        self.save_params.check_key_file();
+
+        // Generate a ed25519 key
+        let mut rng = rand::rngs::StdRng::from_entropy();
+        let ed25519_key = Ed25519PrivateKey::generate(&mut rng);
+
+        // Convert it to the appropriate type and save it
+        let result = match self.key_type {
+            KeyType::X25519 => {
+                let key = x25519::PrivateKey::from_ed25519_private_bytes(&ed25519_key.to_bytes())
+                    .map_err(|err| Error::UnexpectedError(err.to_string()).to_string())?;
+                self.save_params.save_key(&key, "x22519 PrivateKey")
+            }
+            KeyType::Ed25519 => self
+                .save_params
+                .save_key(&ed25519_key, "ed22519 PrivateKey"),
+        };
+
+        to_common_success_result(result)
+    }
+}
+
+#[derive(Debug, StructOpt)]
+pub struct SaveKey {
+    /// Encoding of output key `base64`, `bcs`, or `hex`
+    #[structopt(short, long, default_value = "hex")]
+    encoding: EncodingType,
+    /// Output file
+    #[structopt(short = "f", long, parse(from_os_str))]
+    key_file: PathBuf,
+    /// Force overwrite existing key-file
+    #[structopt(short = "o", long)]
+    overwrite_file: bool,
+}
+
+impl SaveKey {
+    /// Check if the key file exists already
+    pub fn check_key_file(&self) {
+        // Check if file already exists
+        if self.key_file.exists()
+            && !self.overwrite_file
+            && !prompt_yes(
+                format!(
+                    "{:?} already exists, are you sure you want to overwrite it?",
+                    self.key_file.as_os_str()
+                )
+                .as_str(),
+            )
+        {
+            exit(2)
```

Comment:
why exit versus return an error?

Reply:
We can probably return an error, this was just quick to go with.